### PR TITLE
sched/spawn: remove spawn proxy thread to simplify task/posix_spawn()

### PR DESCRIPTION
**Pull request by anchao**
## Summary

sched/spawn: remove spawn proxy thread to simplify task/posix_spawn()

The spawn proxy thread is a special existence in NuttX, usually some developers
spend a lot of time on stack overflow of spawn proxy thread:

https://github.com/apache/nuttx/issues/9046
https://github.com/apache/nuttx/pull/9081

In order to avoid similar issues, this PR will remove spawn thread to simplify
the process of task/posix_spawn().

1. Postpone the related processing of spawn file actions until after task_init()
2. Delete the temporary thread of spawn proxy and related global variables

Signed-off-by: chao an <anchao@xiaomi.com>

## Impact

N/A

## Testing

Image size (sim/nsh):
Patch before:
```bash
$ size nuttx
   text	   data	    bss	    dec	    hex	filename
 397458	  26952	   4064	 428474	  689ba	nuttx
```

Patch after:
```bash
$ size nuttx
   text	   data	    bss	    dec	    hex	filename
 396722	  26936	   4000	 427658	  6868a	nuttx
```

sim/nsh, rv-virt:knsh64

### CODE REVIEW

File: sched/task/task_posixspawn.c [hunk 1]
```diff
@@ -345,91 +269,27 @@ int posix_spawn(FAR pid_t *pid, FAR const char *path,
...
-    }
+      /* Start the task */
 
-  /* Disable pre-emption so that the proxy does not run until waitpid
-   * is called.  This is probably unnecessary since the nxposix_spawn_proxy
-   * has the same priority as this thread; it should be schedule behind
-   * this task in the ready-to-run list.
-   */
+      ret = nxposix_spawn_exec(pid, path, actions, attr, argv, envp);
+#ifdef CONFIG_SCHED_HAVE_PARENT
```
Comment (pussuw) [inline sched/task/task_posixspawn.c:278]: Is this re-parenting necessary any more ? The proxy thread obviously needs to be re-parented but now the thread is created directly from nsh, so the correct parent is nsh right ?

File: binfmt/binfmt_execmodule.c [hunk 2]
```diff
@@ -285,8 +285,14 @@ int exec_module(FAR struct binary_s *binp,
...
 #if defined(CONFIG_ARCH_ADDRENV) || defined(CONFIG_ARCH_VMA_MAPPING)
 errout_with_tcbinit:
-  tcb->cmn.stack_alloc_ptr = NULL;
-  nxsched_release_tcb(&tcb->cmn, TCB_FLAG_TTYPE_TASK);
+#ifndef CONFIG_BUILD_KERNEL
+  if (binp->stackaddr != NULL)
+    {
+      tcb->cmn.stack_alloc_ptr = NULL;
+    }
+#endif
```
Comment (pussuw) [inline binfmt/binfmt_execmodule.c]: Posted question about this in the other PR, as I don't quite understand what was wrong before

File: sched/task/task_spawn.c [hunk 3]
```diff
@@ -131,12 +219,13 @@ static int nxtask_spawn_exec(FAR pid_t *pidp, FAR const char *name,
 
   /* Start the task */
 
-  pid = nxtask_create(name, priority, stackaddr,
-                      stacksize, entry, argv, envp);
+  pid = nxtask_spawn_create(name, priority, stackaddr,
+                            stacksize, entry, argv,
+                            envp ? envp : environ, actions);
```
Comment (pussuw) [inline sched/task/task_spawn.c:224]: If envp = NULL is given from a higher layer, is it correct to grab the parent env here ?

File: sched/task/task_posixspawn.c (hunk 1 above)
Comment (anchao) [reply sched/task/task_posixspawn.c:278]: Yes, you are right, re-parent should not be needed anymore

File: sched/task/task_spawn.c (hunk 3 above)
Comment (anchao) [reply sched/task/task_spawn.c:224]: Yes, it is correct to inherits the parent's env directly

Comment (pussuw) [reply sched/task/task_spawn.c:224]: Ok, thanks

File: include/nuttx/spawn.h [hunk 4]
```diff
@@ -100,6 +100,9 @@ extern "C"
 void add_file_action(FAR posix_spawn_file_actions_t *file_action,
                      FAR struct spawn_general_file_action_s *entry);
 
+int spawn_fileactions(FAR struct tcb_s *tcb,
```
Comment (xiaoxiang781216) [inline include/nuttx/spawn.h:103]: spawn_file_actions

File: sched/task/task_posixspawn.c [hunk 5]
```diff
@@ -345,91 +253,12 @@ int posix_spawn(FAR pid_t *pid, FAR const char *path,
...
-
-errout_with_lock:
-#ifdef CONFIG_SCHED_WAITPID
-  sched_unlock();
-#endif
-  nxmutex_unlock(&g_spawn_parmlock);
-  return ret;
+  return nxposix_spawn_exec(pid, path,
+                            file_actions != NULL ?
+                            *file_actions : NULL, attr, argv, envp);
```
Comment (xiaoxiang781216) [inline sched/task/task_posixspawn.c:263]: why not pass file_actions directly

File: sched/task/task_posixspawn.c [hunk 6]
```diff
@@ -311,30 +239,10 @@ int posix_spawn(FAR pid_t *pid, FAR const char *path,
...
-   */
-
-  if ((file_actions == NULL || *file_actions == NULL) &&
-      (attr == NULL || (attr->flags & POSIX_SPAWN_SETSIGMASK) == 0))
-    {
-      return nxposix_spawn_exec(pid, path, attr, argv, envp);
-    }
-
-  /* Otherwise, we will have to go through an intermediary/proxy task in
+  /* We will have to go through an intermediary/proxy task in
```
Comment (xiaoxiang781216) [inline sched/task/task_posixspawn.c:245]: no intermediary/proxy task anymore?

File: sched/task/task_spawn.c [hunk 7]
```diff
@@ -321,35 +324,13 @@ int task_spawn(FAR const char *name, main_t entry,
...
-      if (ret < 0)
-        {
-          return ret;
-        }
-
-      return pid;
-    }
-
-  /* Otherwise, we will have to go through an intermediary/proxy task in
+  /* We will have to go through an intermediary/proxy task in
```
Comment (xiaoxiang781216) [inline sched/task/task_spawn.c:333]: ditto

File: sched/task/task_spawn.c [hunk 8]
```diff
@@ -360,103 +341,16 @@ int task_spawn(FAR const char *name, main_t entry,
...
     {
-      ret = proxy;
-      serr("ERROR: Failed to start nxtask_spawn_proxy: %d\n", ret);
-      goto errout_with_lock;
+      spawn_proxyattrs(attr);
     }
 
-  /* Wait for the proxy to complete its job */
+  ret = nxtask_spawn_exec(&pid, name, entry,
+                          file_actions != NULL ? *file_actions : NULL,
```
Comment (xiaoxiang781216) [inline sched/task/task_spawn.c:350]: ditto

File: sched/task/task_spawnparms.c [hunk 9]
```diff
@@ -68,76 +58,47 @@ struct spawn_parms_s g_spawn_parms;
...
-          ret = OK;
-        }
-
-      sinfo("Closing fd=%d\n", fd);
-      nx_close(fd);
+      file_close(filep);
     }
 
-  return ret;
+  return file_open(filep, action->path, action->oflags, action->mode);
```
Comment (xiaoxiang781216) [inline sched/task/task_spawnparms.c:101]: let's add nx_open_from_list too

File: include/nuttx/spawn.h (hunk 4 above)
Comment (anchao) [reply include/nuttx/spawn.h:103]: Done

File: sched/task/task_posixspawn.c (hunk 5 above)
Comment (anchao) [reply sched/task/task_posixspawn.c:263]: The prototype of posix_spawn_file_actions_t is a pointer:
https://github.com/apache/nuttx/blob/master/include/spawn.h#L104-L112

File: sched/task/task_posixspawn.c (hunk 6 above)
Comment (anchao) [reply sched/task/task_posixspawn.c:245]: Comments here could be removed

File: sched/task/task_spawn.c (hunk 7 above)
Comment (anchao) [reply sched/task/task_spawn.c:333]: Comments here could be removed

File: sched/task/task_spawnparms.c (hunk 9 above)
Comment (anchao) [reply sched/task/task_spawnparms.c:101]: Done

**Review by xiaoxiang781216 — APPROVED** [review]

**Review by pussuw — APPROVED** [review]
LGTM

**Pull request merged**